Step3

DoTheBestMayB opened the pull request:
pull request 오류는 피드백 반영하며 같이 수정하겠습니다.

DoTheBestMayB committed:
feat(module): app 모듈을 분리

DoTheBestMayB committed:
feat(module): 순수 코틀린으로 변경

DoTheBestMayB committed:
feat(test): test case 작성

DoTheBestMayB committed:
feat(test): test case 세분화

DoTheBestMayB committed:
feat: test case 를 통과하는 코드 작성

DoTheBestMayB committed:
Merge branch 'dothebestmayb' into step3

malibinYun commented on domain/src/main/java/com/github/dodobest/domain/Calculator.kt:
```diff
@@ -0,0 +1,110 @@
+        return inputString.replace(" ", "")
+    }
+
+    private fun splitNumAndSign(inputString: String) : Array<String>{
```
객체지향 생활체조 원칙을 지키면서 프로그래밍을 해보는 것은 어떨까요?
```
1. 한 메서드에 오직 한 단계의 들여쓰기만 한다.
2. else 예약어를 쓰지 않는다.
3. 모든 원시 값과 문자열을 포장한다.
4. 한 줄에 점을 하나만 찍는다.
5. 줄여 쓰지 않는다(축약 금지).
6. 모든 엔티티를 작게 유지한다.
7. 3개 이상의 인스턴스 변수를 가진 클래스를 쓰지 않는다.
8. 일급 컬렉션을 쓴다.
9. getter/setter/프로퍼티를 쓰지 않는다.
```
처음부터 많은 원칙들을 지킬 수는 없지만, 조금씩 지켜나가면 많은 도움이 되실겁니다.
먼저,
1. 한 메서드에 오직 한 단계의 들여쓰기만 한다.
2. else 예약어를 쓰지 않는다.
3. 한 메서드에 10줄을 넘기지 않는다.

이렇게 먼저 연습해보는 것은 어떨까요?

malibinYun commented on domain/src/main/java/com/github/dodobest/domain/Calculator.kt:
```diff
@@ -0,0 +1,110 @@
+        val inputString: String = eraseBlankAtString(input)
+
+        // 숫자와 기호를 나누기
+        val numAndSignArray: Array<String> = splitNumAndSign(inputString)
```
배열 보다는 List 자료구조를 활용해보는 것은 어떨까요?

malibinYun commented on domain/src/main/java/com/github/dodobest/domain/Calculator.kt:
```diff
@@ -0,0 +1,110 @@
+    }
+
+    private fun splitNumAndSign(inputString: String) : Array<String>{
+        val arithmeticOperation: Array<Char> = arrayOf('+', '-', '*', '/') // 사칙연산을 저장한 배열 값
```
코드에 직접 문자열, 문자, 숫자 등을 입력하는 것 보다, 상수화를 시켜 활용하는 것이 좋습니다.
이후 값의 변경에 유연해집니다. 🙂

malibinYun commented on domain/src/main/java/com/github/dodobest/domain/Calculator.kt:
```diff
@@ -0,0 +1,110 @@
+        var firstIdx: Int = -1 // 숫자를 나타내는 처음 Index 값
+        var isMinusSign = false // 숫자가 음수인지 여부를 저장
+
+        for(idx in 0 until inputString.length){
```
하나의 메서드가 너무 많은 일을 하고 있습니다.
기능 명세를 먼저 해보고, 기능 단위로 메서드를 쪼개보는 것은 어떨까요?

malibinYun commented on domain/src/test/java/com/github/dodobest/domain/CalculatorTest.kt:
```diff
@@ -0,0 +1,116 @@
+        val actual: Double = calculator.evaluate("1+2+3")
+
+        // then : 올바른 덧셈 결과를 계산한다
+        assertEquals(6.0, actual, 0.00001)
```
강의에서 배운 Truth를 활용해보는 것은 어떨까요?

malibinYun commented on domain/src/test/java/com/github/dodobest/domain/CalculatorTest.kt:
```diff
@@ -0,0 +1,116 @@
+        ) { calculator.evaluate("120+null/3") }
+
+        // then : IllegalArgumentException 을 발생시킨다
+        assertEquals(IllegalArgumentException::class.java, thrown.javaClass)
```
인스턴스 비교하는 assert 구문을 활용해보는 것은 어떨까요?
```suggestion
        assertThat(thrown).isInstanceOf(IllegalArgumentException::class.java)
```

malibinYun commented on domain/src/test/java/com/github/dodobest/domain/CalculatorTest.kt:
```diff
@@ -0,0 +1,116 @@
+    fun test_exception_input() {
+        // when : 사용자가 예외가 발생하는  사칙연산 식을 입력하면
+        // then : IllegalArgumentException 을 발생시킨다
+        val exceptionStringList: Array<String> = arrayOf("10+ 2 /- 3 * - + 2", "1/0", "2+3*(5-3)", "10+2---2", "10+010")
```
Parameterized Test를 활용해보는 것은 어떨까요?

malibinYun commented on domain/src/test/java/com/github/dodobest/domain/CalculatorTest.kt:
```diff
@@ -0,0 +1,116 @@
+        // then : IllegalArgumentException 을 발생시킨다
+        val exceptionStringList: Array<String> = arrayOf("10+ 2 /- 3 * - + 2", "1/0", "2+3*(5-3)", "10+2---2", "10+010")
+
+        for (exceptionString in exceptionStringList) {
+            println(exceptionStringList.indexOf(exceptionString).toString() + "번 째 " + exceptionString + " test start")
+            val thrown: IllegalArgumentException = assertThrows(
+                IllegalArgumentException::class.java
+            ) { calculator.evaluate(exceptionString) }
+
+            assertEquals(IllegalArgumentException::class.java, thrown.javaClass)
+        }
```
테스트 코드에 로직이 들어가는 것은 지양해야 합니다.
테스트 코드는 프로덕션 코드의 설명서이므로, 눈으로 쭉 읽어 내려갈 수 있는 형태로 작성하는 것이 좋습니다.
for문 같은 로직이 들어가게 되면 테스트 코드를 해석해야 하므로 가독성을 해치는 요소로 작용하게 됩니다.

malibinYun reviewed (changes requested):
3단계 미션 고생하셨습니다. 피드백 반영 후 다시 리뷰 요청 부탁드릴게요! 🙂

DoTheBestMayB committed:
style: 객체지향 생활체조 원칙 준수

DoTheBestMayB committed:
refactor: emptyArray를 List로 변경

emptryArray는 빈 가변 배열로 생각했으나, 빈 불변 배열이다
element를 추가하려고 했는데 append는 되지 않고 += 가능해서 +=을 사용했는데, 이것은 새로운 배열을 생성하는것이라 비효율적임
따라서 List를 사용해야 한다
참고
https://stackoverflow.com/questions/44228815/what-is-the-use-of-emptyarray-in-kotlin
https://stackoverflow.com/questions/33278869/how-do-i-initialize-kotlins-mutablelist-to-empty-mutablelist

DoTheBestMayB committed:
refactor: 사칙연산 기호 상수화

DoTheBestMayB committed:
refactor: Truth 이용하기

DoTheBestMayB committed:
refactor: Parameterized Test 이용하기

참고자료
https://github.com/junit-team/junit4/wiki/parameterized-tests
https://gist.github.com/rossharper/8f6c3c169b6b5c23e12c

DoTheBestMayB committed:
refactor: if else 문을 when 으로 변환

DoTheBestMayB commented:
1. Test코드에서 Parameterized Test를 활용해 봤는데, 올바르게 했는지 궁금합니다.

2. when 조건문에서 class 이름도 is type check가 가능한가요? 관련 자료를 찾지 못해서  `IllegalArgumentException::class.java`의 type을 확인하는 대신 클래스 이름을 확인하는 방법을 사용했습니다.

malibinYun commented on domain/src/main/java/com/github/dodobest/domain/Calculator.kt:
```diff
@@ -0,0 +1,186 @@
+            throw IllegalArgumentException("빈 공백문자를 입력했습니다.")
+        }
+
+        // 문자열로 부터 숫자, 사칙 연산을 분리해서 배열에 추가하기
```
주석으로 의도를 표현하기보다는, 코드에 의도를 녹여내는 방법을 지향해야합니다.
주석을 쓰게 되면 주석 또한 유지 보수의 대상이 됩니다.
하지만 로직이 변경되었을 때 주석에 대한 컴파일 에러 등은 발생하지 않기 때문에, 잊지 않고 유지 보수 하는 것이 매우 어려워집니다. 그러면서 점점 주석은 로직 과는 다른 말을 하게 되고, 결국 거짓말을 하게 되는 주석이 되어버립니다.

주석을 사용하지 않고 최대한 변수와 메서드, 클래스 네이밍을 통해 의도를 나타내려고 연습하는 것을 추천드립니다.

malibinYun commented on domain/src/main/java/com/github/dodobest/domain/Calculator.kt:
```diff
@@ -0,0 +1,186 @@
+class Calculator {
+    companion object {
+        // 사칙 연산 기호
+        const val PLUS: Char = '+'
+        const val MINUS: Char = '-'
+        const val MULTIPLY: Char = '*'
+        const val DIVIDE: Char = '/'
```
"계산기"가 "연산자"의 기호를 가지고 있는 것이 어색합니다.
연산자라는 도메인을 한 곳에 묶어서 클래스의 응집도를 높여보는 것은 어떨까요?

malibinYun replied:
Operator 라는 Enum class 를 만들어 활용해 보세요.

malibinYun commented on domain/src/main/java/com/github/dodobest/domain/Calculator.kt:
```diff
@@ -0,0 +1,186 @@
+        return inputString.replace(" ", "")
+    }
+
+    private fun addNumAndSingToArrayFromString(inputString: String, numAndSignArray: MutableList<String>) {
```
특정 객체 또는 필드를 변형시키는 메서드보다는,
동일한 인풋에 항상 같은 아웃풋을 가지는 메서드로 만드는 것은 어떨까요?

1차 함수인 2x+3 라는 함수가 있다면 x=1 일 때 반드시 5라는 아웃풋이 나오는 것과 같습니다.
위와 같지 않고, 무언가 전역적인 것을 변형시키는 것이 존재한다면, 1이라는 인풋을 넣어도 5뿐만 아니라 다른 값도 반환할 수 있고, 예측 불가능한 값이 나오기도 합니다.

이러한 "순수 함수"는 이런  사이드이팩트를 방지하는 효과도 있습니다.

malibinYun commented on domain/src/main/java/com/github/dodobest/domain/Calculator.kt:
```diff
@@ -0,0 +1,186 @@
+    }
+
+    private fun calcWithOperation(sum: Double, operation: String, num: Double): Double = when (operation) {
+        PLUS.toString() -> sum + num
+        MINUS.toString() -> sum - num
+        MULTIPLY.toString() -> sum * num
+        DIVIDE.toString() -> sum / num
+        else -> {
+            throw IllegalArgumentException("사칙연산 외 기호가 입력되었습니다.")
+        }
```
예외 처리 👍

malibinYun replied:
Enum class를 활용한다면, when 절에 빼먹은 요소가 있을 때 컴파일 에러를 발생 시켜 미처 잡아내지 못한 부분을 고칠 수 있습니다.

malibinYun commented on domain/src/main/java/com/github/dodobest/domain/Calculator.kt:
```diff
@@ -0,0 +1,186 @@
+        return false
+    }
+
+    private fun checkArithmeticOperation(inputString: String, charIndex: Int, numAndSignArray: MutableList<String>) {
```
이 메서드는 좀더 짧게 줄여보는 것은 어떨까요? 🙂

malibinYun commented on domain/src/main/java/com/github/dodobest/domain/Calculator.kt:
```diff
@@ -0,0 +1,186 @@
+        val inputChar: Char = inputString[charIndex]
+
+        // 사칙 연산 기호가 아닌 경우 IllegalArgumentException throw
+        if (!charIsOperation(inputChar)) {
+            throw IllegalArgumentException("사칙 연산 외 기호가 입력되었습니다.")
+        }
```
```suggestion
    require(charIsOperation(inputChar)) { "사칙 연산 외 기호가 입력되었습니다." }
```
코틀린의 require를 활용해볼 수 있습니다.

malibinYun commented on domain/src/test/java/com/github/dodobest/domain/CalculatorTest.kt:
```diff
@@ -0,0 +1,55 @@
+    companion object {
+        @JvmStatic
+        @Parameterized.Parameters
+        fun data() : Collection<Array<Any>> {
```
모든 테스트를 하나의 Parameterized Test를 하는 것은 좋지 못한 방법입니다.
각 테스트에 대해서 어떤 검증을 하고 싶은 지에 대한 의도를 녹여낼 수 없습니다.
또, 테스트는 프로덕션 코드의 설명서 역할을 해야 해서 가독성이 좋아야 하는데, 해당 테스트는 일일이 어떤 설명인지 찾아가며 읽어야 하는 단점이 보입니다.
어떤 테스트가 Parameterized Test가 필요한지, 무엇을 테스트 해야 하는지 나누는 것이 선행되어야 할 것 같습니다.

malibinYun reviewed (approved):
3단계 미션 고생하셨습니다. 피드백은 다음 리뷰에 반영해주세요.
다음 미션도 화이팅입니다!

malibinYun commented:
1. Parameterized Test가 꼭 필요한 부분만 적용해보시면 좋을 것 같습니다.
2. is 예약으로 instance를 비교해볼 수 있지만, 테스트를 먼저 BDD 패턴으로 작성해보시는 것을 추천드립니다.
